lib/internal/loader/Loader.js

```diff
@@ -40,105 +39,83 @@ function normalizeReferrerURL(referrer) {
-    //       -> Promise<{ url : string,
-    //                    format: anything in Loader.validFormats }>
+    //       -> Promise<{ url : string, format: string }>
     // where defaultResolve is ModuleRequest.resolve (having the same
     // signature itself).
     // If `.format` on the returned value is 'dynamic', .dynamicInstantiate
     // will be used as described below.
-    this.resolver = ModuleRequest.resolve;
-    // This hook is only called when resolve(...).format is 'dynamic' and has
-    // the signature
+    this._resolve = defaultResolve;
+    // This hook is only called when resolve(...).format is 'dynamic' and
+    // has the signature
     //   (url : string) -> Promise<{ exports: { ... }, execute: function }>
     // Where `exports` is an object whose property names define the exported
     // names of the generated module. `execute` is a function that receives
     // an object with the same keys as `exports`, whose values are get/set
     // functions for the actual exported values.
-    this.dynamicInstantiate = undefined;
-  }
-
-  hook({ resolve = ModuleRequest.resolve, dynamicInstantiate }) {
-    // Use .bind() to avoid giving access to the Loader instance when it is
-    // called as this.resolver(...);
-    this.resolver = resolve.bind(null);
-    this.dynamicInstantiate = dynamicInstantiate;
+    this._dynamicInstantiate = undefined;
   }
 
-  // Typechecking wrapper around .resolver().
   async resolve(specifier, parentURL = this.base) {
-    if (typeof parentURL !== 'string') {
-      throw new errors.TypeError('ERR_INVALID_ARG_TYPE',
-                                 'parentURL', 'string');
-    }
-
-    const { url, format } = await this.resolver(specifier, parentURL,
-                                                ModuleRequest.resolve);
-
-    if (!Loader.validFormats.includes(format)) {
-      throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'format',
-                                 Loader.validFormats);
-    }
-
-    if (typeof url !== 'string') {
-      throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'url', 'string');
-    }
-
-    if (format === 'builtin') {
-      return { url: `node:${url}`, format };
-    }
+    return this._resolve(specifier, parentURL, defaultResolve);
+  }
 
-    if (format !== 'dynamic') {
-      if (!ModuleRequest.loaders.has(format)) {
-        throw new errors.Error('ERR_UNKNOWN_MODULE_FORMAT', format);
-      }
-      if (!url.startsWith('file:')) {
-        throw new errors.Error('ERR_INVALID_PROTOCOL', url, 'file:');
-      }
-    }
+  async import(specifier, parent = this.base) {
+    const job = await this.getModuleJob(specifier, parent);
+    const module = await job.run();
+    return module.namespace();
+  }
 
-    return { url, format };
+  hook({ resolve, dynamicInstantiate }) {
+    // Use .bind() to avoid giving access to the Loader instance when called.
+    if (resolve !== undefined)
+      this._resolve = resolve.bind(null);
```

Comment:
Should this be made robust against `delete Function.prototype.bind`, or, someone installing their own bind that lets them get access to the Loader instance?